File: Windows_and_Linux/codex_callback_server.py
"""
Codex OAuth Callback Server
----------------------------

A lightweight HTTP server that listens on localhost:1455/auth/callback
to capture the OAuth authorization code from the browser redirect.
Runs in a background thread with a configurable timeout.
"""

import logging
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

from codex_auth import REDIRECT_PORT, CALLBACK_TIMEOUT_SECONDS

_SUCCESS_HTML = """\
<!DOCTYPE html>
<html>
<head>
    <title>Authorization Successful</title>
    <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
            display: flex; justify-content: center; align-items: center;
            height: 100vh; margin: 0;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
        }
        .container {
            background: white; padding: 3rem; border-radius: 1rem;
            box-shadow: 0 10px 40px rgba(0,0,0,0.2); text-align: center;
        }
        h1 { color: #333; margin-bottom: 1rem; }
        p { color: #666; }
    </style>
</head>
<body>
    <div class="container">
        <h1>&#x2713; Authorization Successful</h1>
        <p>You can close this window and return to Writing Tools.</p>
    </div>
    <script>setTimeout(() => window.close(), 2000)</script>
</body>
</html>
"""


class _CallbackHandler(BaseHTTPRequestHandler):
    """HTTP request handler for the OAuth callback."""

    def log_message(self, format, *args):
        # Suppress default stderr logging from BaseHTTPRequestHandler
        logging.debug(f"Callback server: {format % args}")

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path != "/auth/callback":
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"Not found")
            return

        params = parse_qs(parsed.query)

        # Check for error
        if "error" in params:
            error_desc = params.get("error_description", params["error"])[0]
            logging.error(f"OAuth error: {error_desc}")
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"Authorization failed")
            self.server.callback_error = error_desc
            self.server.callback_event.set()
            return

        # Validate state
        received_state = params.get("state", [None])[0]
        if not received_state or received_state != self.server.expected_state:
            logging.error("OAuth state mismatch")
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"Invalid state parameter")
            self.server.callback_error = "State validation failed"
            self.server.callback_event.set()
            return

        # Get code
        code = params.get("code", [None])[0]
        if not code:
            logging.error("Missing authorization code")
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"Missing authorization code")
            self.server.callback_error = "Missing authorization code"
            self.server.callback_event.set()
            return

        # Success — send HTML response and store the code
        self.send_response(200)
        self.send_header("Content-Type", "text/html")
        self.end_headers()
        self.wfile.write(_SUCCESS_HTML.encode("utf-8"))

        self.server.callback_code = code
        self.server.callback_event.set()


class CodexCallbackServer:
    """
    Manages a temporary HTTP server on port 1455 that waits for the OAuth callback.

    Usage:
        server = CodexCallbackServer()
        code = server.wait_for_code(expected_state, timeout=300)
        # code is the authorization code, or None on error/timeout
    """

    def __init__(self):
        self._httpd: HTTPServer | None = None
        self._thread: threading.Thread | None = None

    def wait_for_code(self, expected_state: str, timeout: float = CALLBACK_TIMEOUT_SECONDS) -> str:
        """
        Start the callback server, wait for the browser redirect, return the auth code.
        Raises RuntimeError on timeout, cancellation, or OAuth error.
        """
        event = threading.Event()

        httpd = HTTPServer(("127.0.0.1", REDIRECT_PORT), _CallbackHandler)
        httpd.expected_state = expected_state
        httpd.callback_code = None
        httpd.callback_error = None
        httpd.callback_event = event
        httpd.timeout = 1  # poll interval for handle_request

        self._httpd = httpd

        def serve():
            while not event.is_set():
                httpd.handle_request()

        self._thread = threading.Thread(target=serve, daemon=True)
        self._thread.start()

        logging.debug(f"Callback server listening on port {REDIRECT_PORT}")

        # Wait for the callback (or timeout)
        event.wait(timeout=timeout)

        # Shut down
        self.stop()

        if httpd.callback_error:
            raise RuntimeError(f"OAuth error: {httpd.callback_error}")
        if httpd.callback_code:
            return httpd.callback_code
        raise RuntimeError("OAuth callback timed out")

    def stop(self):
        """Shut down the callback server."""
        if self._httpd:
            try:
                self._httpd.server_close()
            except Exception:
                pass
            self._httpd = None
        self._thread = None

File: macOS/WritingTools/Services/CodexCallbackServer.swift
//
//  CodexCallbackServer.swift
//  WritingTools
//
//  HTTP server for handling OAuth callback from OpenAI
//

import Foundation
import Network

private let logger = AppLogger.logger("CodexCallbackServer")

enum CallbackServerError: LocalizedError {
    case portInUse
    case timeout
    case cancelled
    case invalidState
    case authorizationDenied(String)
    case missingCode
    
    var errorDescription: String? {
        switch self {
        case .portInUse:
            return "Port \(CodexConstants.redirectPort) is already in use. Close other applications using this port."
        case .timeout:
            return "Authorization timed out. Please try again."
        case .cancelled:
            return "Authorization was cancelled"
        case .invalidState:
            return "Invalid OAuth state parameter"
        case .authorizationDenied(let message):
            return "Authorization denied: \(message)"
        case .missingCode:
            return "Authorization code not received"
        }
    }
}

@MainActor
class CodexCallbackServer: ObservableObject {
    @Published var isListening = false
    
    private var listener: NWListener?
    private var expectedState: String?
    private var readyContinuation: CheckedContinuation<Void, Error>?
    private var callbackContinuation: CheckedContinuation<String, Error>?
    private var pendingCallbackResult: Result<String, Error>?
    private var timeoutTask: Task<Void, Never>?
    private var isStarting = false
    
    func startListening(expectedState: String) async throws {
        guard !isListening, !isStarting, listener == nil else {
            throw CallbackServerError.portInUse
        }
        
        self.expectedState = expectedState
        pendingCallbackResult = nil
        isStarting = true
        
        do {
            try setupListener()
            try await withCheckedThrowingContinuation { continuation in
                self.readyContinuation = continuation
            }
            startTimeout()
        } catch {
            resetState()
            throw error
        }
    }

    func waitForCallback() async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            if let pendingCallbackResult {
                self.pendingCallbackResult = nil
                continuation.resume(with: pendingCallbackResult)
            } else {
                self.callbackContinuation = continuation
            }
        }
    }
    
    func stop() {
        logger.debug("Stopping callback server")
        
        let readyContinuation = self.readyContinuation
        self.readyContinuation = nil

        let callbackContinuation = self.callbackContinuation
        self.callbackContinuation = nil

        resetState()

        if let readyContinuation {
            readyContinuation.resume(throwing: CallbackServerError.cancelled)
        }

        if let callbackContinuation {
            callbackContinuation.resume(throwing: CallbackServerError.cancelled)
        }
    }
    
    private func setupListener() throws {
        let parameters = NWParameters.tcp
        parameters.allowLocalEndpointReuse = true
        
        // Enable SO_REUSEADDR and SO_REUSEPORT for better development experience
        let options = parameters.defaultProtocolStack.transportProtocol as? NWProtocolTCP.Options
        options?.enableKeepalive = false
        options?.connectionTimeout = 10
        
        guard let port = NWEndpoint.Port(rawValue: UInt16(CodexConstants.redirectPort)) else {
            throw CallbackServerError.portInUse
        }
        
        do {
            listener = try NWListener(using: parameters, on: port)
        } catch let error as NSError {
            // Provide more detailed error message
            logger.error("Failed to create listener on port \(CodexConstants.redirectPort): \(error.localizedDescription)")
            if error.domain == NSPOSIXErrorDomain && error.code == Int(EADDRINUSE) {
                logger.error("Port \(CodexConstants.redirectPort) is already in use. Close other WritingTools instances or apps using this port.")
            }
            throw CallbackServerError.portInUse
        }
        
        listener?.stateUpdateHandler = { [weak self] state in
            Task { @MainActor in
                self?.handleStateChange(state)
            }
        }
        
        listener?.newConnectionHandler = { [weak self] connection in
            Task { @MainActor in
                self?.handleConnection(connection)
            }
        }
        
        listener?.start(queue: .main)
        logger.debug("Callback server started on port \(CodexConstants.redirectPort)")
    }
    
    private func handleStateChange(_ state: NWListener.State) {
        switch state {
        case .ready:
            logger.debug("Listener ready")
            isStarting = false
            isListening = true
            readyContinuation?.resume()
            readyContinuation = nil
        case .failed(let error):
            logger.error("Listener failed: \(error.localizedDescription)")
            if let readyContinuation {
                self.readyContinuation = nil
                resetState()
                readyContinuation.resume(throwing: CallbackServerError.portInUse)
            } else {
                completeWithError(CallbackServerError.portInUse)
            }
        case .cancelled:
            logger.debug("Listener cancelled")
        default:
            break
        }
    }
    
    private func handleConnection(_ connection: NWConnection) {
        connection.stateUpdateHandler = { state in
            if case .failed(let error) = state {
                logger.warning("Connection failed: \(error.localizedDescription)")
            }
        }
        
        connection.start(queue: .main)
        
        receiveRequest(from: connection)
    }
    
    private func receiveRequest(from connection: NWConnection) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
            Task { @MainActor in
                guard let self = self else { return }
                
                if let error = error {
                    logger.error("Receive error: \(error.localizedDescription)")
                    self.sendResponse(to: connection, statusCode: 500, body: "Internal error")
                    return
                }
                
                guard let data = data, let requestString = String(data: data, encoding: .utf8) else {
                    self.sendResponse(to: connection, statusCode: 400, body: "Invalid request")
                    return
                }
                
                self.handleHTTPRequest(requestString, connection: connection)
            }
        }
    }
    
    private func handleHTTPRequest(_ requestString: String, connection: NWConnection) {
        logger.debug("Received HTTP request")
        
        // Parse the request line
        guard let firstLine = requestString.split(separator: "\r\n").first,
              let urlPart = firstLine.split(separator: " ").dropFirst().first else {
            sendResponse(to: connection, statusCode: 400, body: "Invalid request format")
            return
        }
        
        // Parse query parameters
        let urlString = String(urlPart)
        guard let components = URLComponents(string: "http://localhost\(urlString)") else {
            sendResponse(to: connection, statusCode: 400, body: "Invalid URL")
            return
        }
        
        // Check if this is the callback path
        guard components.path == "/auth/callback" else {
            sendResponse(to: connection, statusCode: 404, body: "Not found")
            return
        }
        
        let queryItems = components.queryItems ?? []
        let params = Dictionary(uniqueKeysWithValues: queryItems.map { ($0.name, $0.value ?? "") })
        
        // Check for error
        if let error = params["error"] {
            let errorDescription = params["error_description"] ?? error
            logger.error("OAuth error: \(errorDescription)")
            sendResponse(to: connection, statusCode: 400, body: "Authorization failed")
            completeWithError(CallbackServerError.authorizationDenied(errorDescription))
            return
        }
        
        // Validate state
        guard let receivedState = params["state"], !receivedState.isEmpty else {
            logger.error("Missing state parameter")
            sendResponse(to: connection, statusCode: 400, body: "Missing state parameter")
            completeWithError(CallbackServerError.invalidState)
            return
        }
        
        guard receivedState == self.expectedState else {
            let expected = self.expectedState ?? "nil"
            logger.error("State mismatch: expected \(expected), got \(receivedState)")
            sendResponse(to: connection, statusCode: 400, body: "Invalid state parameter")
            completeWithError(CallbackServerError.invalidState)
            return
        }
        
        // Get authorization code
        guard let code = params["code"], !code.isEmpty else {
            logger.error("Missing authorization code")
            sendResponse(to: connection, statusCode: 400, body: "Missing authorization code")
            completeWithError(CallbackServerError.missingCode)
            return
        }
        
        logger.debug("Successfully received authorization code")
        sendSuccessResponse(to: connection)
        completeWithCode(code)
    }
    
    private func sendResponse(to connection: NWConnection, statusCode: Int, body: String) {
        let response = """
        HTTP/1.1 \(statusCode) \(httpStatusText(statusCode))
        Content-Type: text/plain
        Content-Length: \(body.utf8.count)
        Connection: close
        
        \(body)
        """
        
        guard let data = response.data(using: .utf8) else { return }
        
        connection.send(content: data, completion: .contentProcessed { _ in
            connection.cancel()
        })
    }
    
    private func sendSuccessResponse(to connection: NWConnection) {
        let html = """
        <!DOCTYPE html>
        <html>
        <head>
            <title>Authorization Successful</title>
            <style>
                body {
                    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    height: 100vh;
                    margin: 0;
                    background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
                }
                .container {
                    background: white;
                    padding: 3rem;
                    border-radius: 1rem;
                    box-shadow: 0 10px 40px rgba(0,0,0,0.2);
                    text-align: center;
                }
                h1 { color: #333; margin-bottom: 1rem; }
                p { color: #666; }
            </style>
        </head>
        <body>
            <div class="container">
                <h1>✓ Authorization Successful</h1>
                <p>You can close this window and return to Writing Tools.</p>
            </div>
            <script>setTimeout(() => window.close(), 2000)</script>
        </body>
        </html>
        """
        
        let response = """
        HTTP/1.1 200 OK
        Content-Type: text/html
        Content-Length: \(html.utf8.count)
        Connection: close
        
        \(html)
        """
        
        guard let data = response.data(using: .utf8) else { return }
        
        connection.send(content: data, completion: .contentProcessed { _ in
            connection.cancel()
        })
    }
    
    private func httpStatusText(_ code: Int) -> String {
        switch code {
        case 200: return "OK"
        case 400: return "Bad Request"
        case 404: return "Not Found"
        case 500: return "Internal Server Error"
        default: return "Unknown"
        }
    }
    
    private func startTimeout() {
        timeoutTask = Task {
            try? await Task.sleep(for: .seconds(CodexConstants.callbackTimeoutSeconds))
            
            if !Task.isCancelled {
                await MainActor.run {
                    logger.warning("Callback server timed out")
                    completeWithError(CallbackServerError.timeout)
                }
            }
        }
    }
    
    private func completeWithCode(_ code: String) {
        timeoutTask?.cancel()
        timeoutTask = nil
        
        if let callbackContinuation {
            self.callbackContinuation = nil
            callbackContinuation.resume(returning: code)
        } else {
            pendingCallbackResult = .success(code)
        }

        scheduleStop()
    }
    
    private func completeWithError(_ error: Error) {
        timeoutTask?.cancel()
        timeoutTask = nil
        
        if let callbackContinuation {
            self.callbackContinuation = nil
            callbackContinuation.resume(throwing: error)
        } else {
            pendingCallbackResult = .failure(error)
        }

        scheduleStop()
    }

    private func scheduleStop() {
        Task {
            try? await Task.sleep(for: .milliseconds(500))
            await MainActor.run {
                stop()
            }
        }
    }

    private func resetState() {
        timeoutTask?.cancel()
        timeoutTask = nil
        listener?.cancel()
        listener = nil
        isListening = false
        isStarting = false
        expectedState = nil
    }
    
    deinit {
        listener?.cancel()
        timeoutTask?.cancel()
    }
}
